### PR TITLE
wip: add ts support

### PR DESCRIPTION
**Pull request by z4o4z**
this PR adds support for TS(#26), it generates normals JS build but with typescript definitions

closes #25

### CODE REVIEW

**Comment by IljaDaderko** [comment]
Could you please add changes from https://github.com/alewin/useWorker/pull/25 directly into new typescript source so that async function returned by `useWorker` inherits type of it's original.

**Comment by z4o4z** [comment]
> Could you please add changes from #25 directly into new typescript source so that async function returned by `useWorker` inherits type of its original. That PR won't be necessary after these changes.

yeah sure, also I've fixed some types issue

File: src/lib/jobRunner.ts [hunk 1]
```diff
@@ -11,15 +11,15 @@
  * @returns {Function} returns a function that accepts the parameters
  * to be passed to the "userFunc" function
  */
-const jobRunner = userFunc => e => {
+const jobRunner = (userFunc: Function) => (e: { data: any[] }) => {
```
Comment (grdnrt) [inline src/lib/jobRunner.ts]: What do you guys think about using `e: MessageEvent` here instead? Per [MDN](https://developer.mozilla.org/en-US/docs/Web/API/Worker) cc @alewin @IljaDaderko

Comment (IljaDaderko) [reply src/lib/jobRunner.ts]: Makes sense to me 👍

Comment (grdnrt) [reply src/lib/jobRunner.ts]: Oh cool, seems that OP just missed it here :)

Comment (alewin) [reply src/lib/jobRunner.ts]: yes, it seems right! 👌

Comment (z4o4z) [reply src/lib/jobRunner.ts]: @grdnrt @alewin @IljaDaderko updated, could someone take a look ?)

File: src/useWorker.ts [hunk 2]
```diff
@@ -0,0 +1,104 @@
+/* eslint-disable no-underscore-dangle */
+/* eslint-disable no-console */
```
Comment (IljaDaderko) [inline src/useWorker.ts]:
```suggestion
/* eslint-disable no-underscore-dangle, no-console */
```

Or we can disable these rules globally?

File: src/useWorker.ts [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+import React from 'react'
+import createWorkerBlobUrl from './lib/createWorkerBlobUrl'
+import WORKER_STATUS from './lib/status'
+
+type Options = {
+  timeout?: number;
+  dependencies: string[];
```
Comment (IljaDaderko) [inline src/useWorker.ts]: I think dependencies are optional as well.

File: src/useWorker.ts [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+ * @param {Object} options useWorker option params
+ */
+export const useWorker = <T extends (...fnArgs: any[]) => any>(fn: T, options: Options = DEFAULT_OPTIONS) => {
+  const [workerStatus, setWorkerStatus] = React.useState<WORKER_STATUS>(WORKER_STATUS.PENDING)
+  const worker = React.useRef<Worker & { _url?: string }>()
+  const promise = React.useRef<{ [PROMISE_REJECT]?: (result: ReturnType<T> | ErrorEvent) => void;[PROMISE_RESOLVE]?: (result: ReturnType<T>) => void }>({})
+  const timeoutId = React.useRef<number>()
+
+  const killWorker = (status = WORKER_STATUS.PENDING) => {
+    if (worker.current && worker.current._url) {
```
Comment (IljaDaderko) [inline src/useWorker.ts]:
```suggestion
    if (worker.current?._url) {
```

File: src/useWorker.ts [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+    newWorker.onerror = (e: ErrorEvent) => {
+      promise.current[PROMISE_REJECT]!(e)
+      killWorker(WORKER_STATUS.ERROR)
+    }
+
+    if (timeout) {
+      timeoutId.current = setTimeout(() => {
+        killWorker(WORKER_STATUS.TIMEOUT_EXPIRED)
+      }, timeout) as any as number
```
Comment (IljaDaderko) [inline src/useWorker.ts]: `as any as number` I never seen 2 `as` used like this before, out of curiosity what does this do?

File: src/useWorker.ts (hunk 2 above)
Comment (alewin) [reply src/useWorker.ts]: Yes after merging this PR, I want to open a PR removing these comments, and rearranging the package.json (remove babel6, prettier etc.) 👍

File: src/useWorker.ts [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+  timeout: undefined,
+  dependencies: [],
+}
+
+/**
+ *
+ * @param {Function} fn the function to run with web worker
+ * @param {Object} options useWorker option params
+ */
+export const useWorker = <T extends (...fnArgs: any[]) => any>(fn: T, options: Options = DEFAULT_OPTIONS) => {
```
Comment (alewin) [inline src/useWorker.ts]: `options` should be optional `options?: Options`, or isn't it necessary since I have a default?

File: src/useWorker.ts [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+  const [workerStatus, setWorkerStatus] = React.useState<WORKER_STATUS>(WORKER_STATUS.PENDING)
+  const worker = React.useRef<Worker & { _url?: string }>()
+  const promise = React.useRef<{ [PROMISE_REJECT]?: (result: ReturnType<T> | ErrorEvent) => void;[PROMISE_RESOLVE]?: (result: ReturnType<T>) => void }>({})
+  const timeoutId = React.useRef<number>()
```
Comment (Pigotz) [inline src/useWorker.ts]:
```suggestion
  const timeoutId = React.useRef<ReturnType<typeof setTimeout>>()
```

This should be safer

File: src/useWorker.ts (hunk 5 above)
Comment (Pigotz) [reply src/useWorker.ts]: I just made a suggestion above, that should remove the need of this cast

File: src/useWorker.ts (hunk 4 above)
Comment (z4o4z) [reply src/useWorker.ts]: the build tool doesn't support optional chaining with TS :(

File: src/useWorker.ts (hunk 5 above)
Comment (z4o4z) [reply src/useWorker.ts]: the reason for doing that is that "@types/node" is one of the dependencies, and "setTimeout" returns not a number in the Node, I've found an appropriate solution for this type stuff

File: src/useWorker.ts (hunk 4 above)
Comment (alewin) [reply src/useWorker.ts]: We can install `@babel/plugin-proposal-optional-chaining` and adding the plugin into `.babelrc`, It should work

**.babelrc**
```json
{
    "plugins": [
        [
            "@babel/plugin-proposal-optional-chaining",
            {
                "loose": false
            }
        ]
    ]
}
```

I can take care of it, once we merged this PR 👍

cc: @IljaDaderko @z4o4z

File: src/useWorker.ts (hunk 5 above)
Comment (alewin) [reply src/useWorker.ts]: Resolved adding `window.setTimeout`

More info: https://stackoverflow.com/a/51040768/1477950

**Comment by IljaDaderko** [comment]
I think that besides making options optional this looks ready for test drive, perhaps we can make rc?

File: package.json [hunk 8]
```diff
@@ -1,20 +1,21 @@
...
-  "types": "./index.d.ts",
+  "types": "dist/index.d.ts",
   "source": "src/index.js",
   "main": "dist/index.js",
   "umd:main": "dist/index.umd.js",
   "module": "dist/index.mjs",
   "scripts": {
-    "build": "microbundle src/index.js",
+    "build": "microbundle src/index.ts",
     "dev": "microbundle watch",
```
Comment (alewin) [inline package.json]: `dev` command doesn't work @z4o4z , we need to:

- change `"source": "src/index.js"` into `"source": "src/index.ts"`

- add into `.gitignore` the generated cache directories

.gitignore:
```
.rts2_cache_*
```

**Comment by alewin** [comment]
> I think that besides making options optional this looks ready for test drive, perhaps we can make rc?

Yes there is also to fix the command `dev`, and then we can finally make rc :)

File: package.json (hunk 8 above)
Comment (z4o4z) [reply package.json]: fixed
cc @alewin @IljaDaderko

File: src/useWorker.ts (hunk 6 above)
Comment (alewin) [reply src/useWorker.ts]: I did a test and define `options?: Options` as optional isn't necessary since we have `DEFAULT_OPTIONS`

**Comment by alewin** [comment]
Before making the RC I did a test, and there is a problem with the `postmessage` function used inside the `webworker` ( **jobRunner.ts** )

**Problem**
---

Adding `*` as "targetOrigin " causes a runtime-error 
```javascript
postMessage(['SUCCESS', result], '*')
```
https://github.com/alewin/useWorker/pull/27/files#diff-b9c4a8a5ca6a13d205a660aaea26ed67R19

---

**Possibile Solutions:**:

- https://stackoverflow.com/questions/48612304/running-inlined-webworker-code-from-within-typescript-project 

adding  `@ts-ignore` before "postMessage" seems works, but its not very clean.
What do you think 🤔?

**Comment by z4o4z** [comment]
> Before making the RC I did a test, and there is a problem with the `postmessage` function used inside the `webworker` ( **jobRunner.ts** )
> 
> ## **Problem**
> Adding `*` as "targetOrigin " causes a runtime-error
> 
> ```js
> postMessage(['SUCCESS', result], '*')
> ```
> 
> https://github.com/alewin/useWorker/pull/27/files#diff-b9c4a8a5ca6a13d205a660aaea26ed67R19
> 
> **Possibile Solutions:**:
> 
> * https://stackoverflow.com/questions/48612304/running-inlined-webworker-code-from-within-typescript-project
> 
> adding `@ts-ignore` before "postMessage" seems works, but its not very clean.
> What do you think 🤔?
> 
> **More info**
> Issue: [microsoft/TypeScript#30042](https://github.com/microsoft/TypeScript/issues/30042) Pull Request: [microsoft/TSJS-lib-generator#672](https://github.com/microsoft/TSJS-lib-generator/pull/672)

this is a really good find! 
I guess `@ts-ignore` is OK for now

**Comment by alewin** [comment]
Great! Thanks everyone for your cooperation! 😃
@z4o4z @IljaDaderko @grdnrt @Pigotz

**Pull request merged**